File: app/src/main/java/com/wavesplatform/wallet/v1/crypto/PublicKeyAccount.java
package com.wavesplatform.wallet.v1.crypto;

import com.wavesplatform.wallet.v1.util.AddressUtil;

public class PublicKeyAccount {
    public static class InvalidPublicKey extends Exception {}

    private final byte[] publicKey;
    private final String address;
    private final String publicKeyStr;

    public static int KeyStringLength = Base58.base58Length(32);

    public PublicKeyAccount(String s) throws InvalidPublicKey {
        this.publicKeyStr = s;
        if (s.length() > KeyStringLength) throw new InvalidPublicKey();
        try {
            this.publicKey = Base58.decode(s);
        } catch (Base58.InvalidBase58 invalidBase58) {
            throw new InvalidPublicKey();
        }
        this.address = AddressUtil.addressFromPublicKey(publicKey);
    }

    public byte[] getPublicKey() {
        return publicKey;
    }

    public String getAddress() {
        return address;
    }

    public String getPublicKeyStr() {
        return publicKeyStr;
    }

}

File: wavesplatform/src/main/java/com/wavesplatform/sdk/crypto/PublicKeyAccount.kt
/*
 * Created by Eduard Zaydel on 1/4/2019
 * Copyright © 2019 Waves Platform. All rights reserved.
 */

package com.wavesplatform.sdk.crypto

import com.wavesplatform.sdk.utils.addressFromPublicKey

class PublicKeyAccount @Throws(InvalidPublicKey::class)
constructor(val publicKeyStr: String) {

    val publicKey: ByteArray
    val address: String

    class InvalidPublicKey : Exception()

    init {
        if (publicKeyStr.length > KeyStringLength) throw InvalidPublicKey()
        try {
            this.publicKey = Base58.decode(publicKeyStr)
        } catch (invalidBase58: Base58.InvalidBase58) {
            throw InvalidPublicKey()
        }

        this.address = addressFromPublicKey(publicKey)
    }

    companion object {
        var KeyStringLength = Base58.base58Length(32)
    }
}
